this project is the binary classification problem between two varabled Mine and Rock on the submarien which detect rock and mine on water as it scans were it going  

In [7]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score

In [8]:
file_path =r'C:\Users\Wsiwale\Desktop\Dataset\sonar data.csv'
sonar_data = pd.read_csv(file_path)

In [9]:
sonar_data.head()

0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109  \
0  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
1  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
2  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
3  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
4  0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   

   0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084  \
0  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
1  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
2  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
3  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
4  0.3039  ...  0.0045  0.0014  0.0038  0.0013  0.0089  0.0057  0.0027   

   0.0090  0.0032  R  
0  0.0052  0.0044  R  
1  0.0095  0.0078  R  
2  0.0040  0.0117  R  
3  0.0107  0.0094  R  
4  0.0051  0.0062  R  

[5 rows x 61 columns]

In [10]:
sonar_data.describe()

0.0200      0.0371      0.0428      0.0207      0.0954      0.0986  \
count  207.000000  207.000000  207.000000  207.000000  207.000000  207.000000   
mean     0.029208    0.038443    0.043837    0.054053    0.075105    0.104599   
std      0.023038    0.033040    0.038521    0.046583    0.055669    0.059247   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013300    0.016400    0.018900    0.024450    0.037700    0.066950   
50%      0.022800    0.030800    0.034200    0.044100    0.062000    0.092100   
75%      0.035800    0.048100    0.058200    0.065700    0.101050    0.134150   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

           0.1539      0.1601      0.3109      0.2111  ...      0.0232  \
count  207.000000  207.000000  207.000000  207.000000  ...  207.000000   
mean     0.121591    0.134677    0.177361    0.208245  ...    0.016034   
std      0.061897    0.085340    0.118311    0.134741  ...    0.012027   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080600    0.080350    0.096750    0.111150  ...    0.008350   
50%      0.105600    0.111900    0.152200    0.181000  ...    0.013800   
75%      0.153050    0.169800    0.231500    0.269000  ...    0.020700   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

           0.0027      0.0065      0.0159      0.0072      0.0167      0.0180  \
count  207.000000  207.000000  207.000000  207.000000  207.000000  207.000000   
mean     0.013472    0.010729    0.010917    0.009300    0.008181    0.007771   
std      0.009628    0.007071    0.007310    0.007103    0.005719    0.005756   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007350    0.005050    0.005350    0.004100    0.004400    0.003700   
50%      0.011500    0.009600    0.009300    0.007500    0.006800    0.005900   
75%      0.016750    0.014900    0.014450    0.012100    0.010350    0.010350   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

           0.0084      0.0090      0.0032  
count  207.000000  207.000000  207.000000  
mean     0.007947    0.007936    0.006523  
std      0.006485    0.006196    0.005038  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003650    0.003100  
50%      0.005800    0.006300    0.005300  
75%      0.010400    0.010350    0.008550  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [11]:
sonar_data.shape

(207, 61)

how many samples are their for each between rock and mine form the beging to the end of the data removing the results if given for a traning set in this case is 60 the 61 is the colonm for results

In [12]:
sonar_data.iloc[:, 60].value_counts()

R
M    111
R     96
Name: count, dtype: int64

.iloc[:, 60]
.iloc[] is used to select rows and columns from a pandas DataFrame by their integer position (i.e., index-based selection).
[:, 60] specifies that:
: means "select all rows."
60 refers to the 61st column (since indexing starts at 0).
So, sonar_data.iloc[:, 60] selects all rows from the 61st column in the DataFrame.

In [13]:
print(sonar_data.columns)

Index(['0.0200', '0.0371', '0.0428', '0.0207', '0.0954', '0.0986', '0.1539',
       '0.1601', '0.3109', '0.2111', '0.1609', '0.1582', '0.2238', '0.0645',
       '0.0660', '0.2273', '0.3100', '0.2999', '0.5078', '0.4797', '0.5783',
       '0.5071', '0.4328', '0.5550', '0.6711', '0.6415', '0.7104', '0.8080',
       '0.6791', '0.3857', '0.1307', '0.2604', '0.5121', '0.7547', '0.8537',
       '0.8507', '0.6692', '0.6097', '0.4943', '0.2744', '0.0510', '0.2834',
       '0.2825', '0.4256', '0.2641', '0.1386', '0.1051', '0.1343', '0.0383',
       '0.0324', '0.0232', '0.0027', '0.0065', '0.0159', '0.0072', '0.0167',
       '0.0180', '0.0084', '0.0090', '0.0032', 'R'],
      dtype='object')


In [14]:
print(sonar_data.shape)

(207, 61)


In [15]:
header =list(range(61))
sonar_data = pd.read_csv(file_path, header=None, names=header)

In [16]:
sonar_data.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [17]:
sonar_data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [18]:
header =list(range(61))
sonar_data = pd.read_csv(file_path, header=None, names=header)

In [19]:
sonar_data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

now we are going to separating data and labels

In [20]:
X = sonar_data.drop(columns=60, axis=1) 
Y = sonar_data[60]

Explanation:
drop(columns=60, axis=1) correctly specifies that you want to drop the column with index 60.
columns=60: This tells pandas to drop column 60.
axis=1: Specifies that we are referring to columns (not rows).

In [21]:
print(X)
print(Y)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

Training and test data

Explanation:
train_test_split(X, Y, ...):

This function is used to split the dataset into two parts:
Training set: Used to train the machine learning model.
Testing set: Used to evaluate the model’s performance on unseen data.
Arguments:

X: The features (input variables) of the dataset (all columns except for column 60, as per the earlier step).
Y: The target (output) variable, which is column 60 (for example, class labels like "M" for mines or "R" for rocks).
Both X and Y will be split into corresponding training and testing sets.

test_size=0.1:

This specifies that 10% of the data will be used for the test set, while the remaining 90% will be used for the training set.
stratify=Y:

This ensures that the class distribution in Y (target variable) is preserved in both the training and test sets.
For example, if 70% of your labels are class M (mines) and 30% are class R (rocks), then both the training and test sets will maintain approximately the same distribution.
random_state=1:

This sets a seed for the random number generator, ensuring that the split will be the same every time you run the code. It ensures reproducibility of the results.
Output:
X_train: The training set for the features (90% of the original X).
X_test: The test set for the features (10% of the original X).
Y_train: The training set for the target variable (90% of the original Y).
Y_test: The test set for the target variable (10% of the original Y).

In [22]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.1, stratify = Y, random_state = 1)

In [23]:
print(X.shape, X_train.shape, X_test.shape)

(208, 60) (187, 60) (21, 60)


Explanation:
X.shape:

This gives the shape of the features (input variables) before the split.
The output will be in the form (number_of_rows, number_of_columns), representing the total number of samples and features in the dataset.
X_train.shape:

This gives the shape of the training set after the split.
Since the training set contains 90% of the data (due to test_size=0.1), the first number will be approximately 90% of the original rows from X.
X_test.shape:

This gives the shape of the test set after the split.
Since the test set contains 10% of the data, the first number will be approximately 10% of the original rows from X.

out:
(208, 60): The original dataset (X) has 208 rows (samples) and 60 columns (features).
(187, 60): After splitting, the training set (X_train) contains 187 rows, which is 90% of 208.
(21, 60): The test set (X_test) contains 21 rows, which is 10% of 208.

model training 

In [24]:
print(X_train)
print(Y_train)

         0       1       2       3       4       5       6       7       8   \
115  0.0414  0.0436  0.0447  0.0844  0.0419  0.1215  0.2002  0.1516  0.0818   
38   0.0123  0.0022  0.0196  0.0206  0.0180  0.0492  0.0033  0.0398  0.0791   
56   0.0152  0.0102  0.0113  0.0263  0.0097  0.0391  0.0857  0.0915  0.0949   
123  0.0270  0.0163  0.0341  0.0247  0.0822  0.1256  0.1323  0.1584  0.2017   
18   0.0270  0.0092  0.0145  0.0278  0.0412  0.0757  0.1026  0.1138  0.0794   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
140  0.0412  0.1135  0.0518  0.0232  0.0646  0.1124  0.1787  0.2407  0.2682   
5    0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   
154  0.0117  0.0069  0.0279  0.0583  0.0915  0.1267  0.1577  0.1927  0.2361   
131  0.1150  0.1163  0.0866  0.0358  0.0232  0.1267  0.2417  0.2661  0.4346   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   

         9   ...      50      51      52      53   

In [25]:
model = LogisticRegression()

In [26]:
model.fit(X_train, Y_train)

LogisticRegression()

Accuracy on training data

In [27]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [28]:
print("Accuracy on Training Data : ", training_data_accuracy)

Accuracy on Training Data :  0.8342245989304813


accuracy on test data

In [29]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [33]:
print("Accuracy on Test Data : ", test_data_accuracy)

Accuracy on Test Data :  0.7619047619047619


Making a prediction system

In [35]:
input_data =(0.0442, 0.0477, 0.0049, 0.0581, 0.0278, 0.0678, 0.1664, 0.149,	0.0974,	0.1268,	0.1109,	0.2375, 0.2007,	0.214,	0.1109,	0.2036,	0.2468,	0.6682,	0.8345,	0.8252,	0.8017,	0.8982,	0.9664,	0.8515,	0.6626,	0.3241,	0.2054,	0.5669,	0.5726,	0.4877,	0.7532,	0.76,	0.5185,	0.412,	0.556,	0.5569,	0.1336,	0.3831,	0.4611,	0.433,	0.2556,	0.1466,	0.3489,	0.2659,	0.0944,	0.137,	0.1344,	0.0416,	0.0719,	0.0637,	0.021,	0.0204,	0.0216,	0.0135,	0.0055,	0.0073,	0.008,	0.0105,	0.0059,	0.0105
)
# chaing the input data to numpy array
input_data_as_numpy_array = np.asarray(input_data)
# reshape the numpy array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)
prediction = model.predict(input_data_reshaped)
print(prediction)

if(prediction[0]== "M"):
    print("The object is a Mine")
else:
    print("The Object is a Rock")

['M']
The object is a Mine


Here's a step-by-step explanation of the code you provided:

1. Define the Input Data:
python
Copy code
input_data = (0.0442, 0.0477, 0.0049, 0.0581, 0.0278, 0.0678, 0.1664, 0.149, 0.0974, 0.1268, 0.1109, 0.2375, 0.2007, 0.214, 0.1109, 0.2036, 0.2468, 0.6682, 0.8345, 0.8252, 0.8017, 0.8982, 0.9664, 0.8515, 0.6626, 0.3241, 0.2054, 0.5669, 0.5726, 0.4877, 0.7532, 0.76, 0.5185, 0.412, 0.556, 0.5569, 0.1336, 0.3831, 0.4611, 0.433, 0.2556, 0.1466, 0.3489, 0.2659, 0.0944, 0.137, 0.1344, 0.0416, 0.0719, 0.0637, 0.021, 0.0204, 0.0216, 0.0135, 0.0055, 0.0073, 0.008, 0.0105, 0.0059, 0.0105)
This defines a tuple named input_data that contains 58 numerical values. These values represent a single data instance, possibly features for machine learning prediction (e.g., in a classification or regression task).

2. Convert the Tuple to a NumPy Array:
python
Copy code
input_data_as_numpy_array = np.asarray(input_data)
Here, input_data is converted from a Python tuple to a NumPy array using np.asarray(). NumPy arrays are a preferred format for numerical computation in machine learning because they are optimized for fast calculations.

Reason: Machine learning models often require inputs to be in NumPy array format for computation.
3. Reshape the Data:
python
Copy code
input_data_reshaped = input_data_as_numpy_array.reshape(1, -1)
The reshape(1, -1) reshapes the NumPy array into a 2D array with 1 row and 58 columns. This is necessary because most machine learning models expect a 2D input where:

Each row represents an instance (in this case, one instance).

Each column represents a feature of the instance.

Explanation: reshape(1, -1) ensures the array has the shape (1, 58), where 1 represents the number of instances (one in this case) and 58 represents the number of features.

4. Make a Prediction:
python
Copy code
prediction = model.predict(input_data_reshaped)
This line uses a pre-trained machine learning model (model) to make a prediction on the reshaped input data. The model applies the learned patterns to the input data and returns a prediction.

Explanation: model.predict(input_data_reshaped) passes the reshaped input data into the model's predict() function, which outputs a prediction (classification label, probability, or continuous value depending on the model type).
Note: For this to work, you must have a model (model) already trained or loaded. The code assumes that model has been defined earlier in the script.

5. Print the Prediction:
python
Copy code
print(prediction)
Finally, this line prints the output of the prediction. The format and type of the output depend on the model:

If it's a classification model, it might return a label (e.g., 0 or 1) or probabilities.
If it's a regression model, it will return a continuous value (e.g., predicted price, score, etc.).
Summary:
This code prepares a single instance of input data for making a prediction using a machine learning model. It:

Converts the data into a NumPy array.
Reshapes it to match the expected input shape of the model.
Passes the reshaped data to a trained model for prediction.
Prints the resulting prediction.
For the code to work, you need to ensure that the model is defined or loaded, and the required libraries such as NumPy are imported (import numpy as np).